decisions/DR008-Unify-the-entity-data-model.md

```diff
@@ -0,0 +1,358 @@
+project.
+
+We believe these benefits outweigh the specific case of generic
+resolution expressed through a first-class API call, as this can be
+recreated with a suitably generic `string` trait if specifically
+required.
+
+In addition, for a significant number of projected use cases, multiple
+API calls (to resolve then attributes) are needed anyway. The combined
+proposal provides for more optimization opportunities for these cases
+than then existing API.
+
+The resulting demands made on the manager's implementation are
+approximately the same in either case as there are always entities that
+were authored outside of OpenAssetIO interactions, that would require
+work to determine what a suitable primary string (or string trait value)
+should be.
+
+We believe that in real-world scenarios, any overhead the data
+structures required to support the trait-based return types would be
+relatively insignificant compared to all other work performed by a
+manager to satisfy any given lookup.
+
+### Performance measurements
+
+A valid concern regarding this change is that it precludes a happy-path
+optimization were a single string of a single trait is required for
+resolution, and the overhead of a view-based API would be significant.
+
+The following timings were obtained using a naive, un-optimized C++
+implementation of the [generic resolution](#generic-resolution) sketch
+in the appendix below. In the implementation, the manager is in the
+ideal situation where results are cached in-process in a pre-populated
+`std::unordered_map`:
+
+```
+Queries: 100000
+Batch size: 1
+Wall-clock time (Total):
+  resolveEntityReference: 43 ms
+  resolve: 124 ms
+Wall-clock time (Per-call):
+  resolveEntityReference: 0.00043ms
+  resolve: 0.00124ms
+```
+
+We believe that in typical real-world scenarios, adding less than tenth
+of a second onto the time to resolve a _hundred thousand_ entities is
+not a significant concern compared to the rest of work-to-be-done by the
+system to satisfy those queries, or make use of the resulting data.
+
+## Appendix
+
+What follows in a comparative illustration of the two approaches, the
+code for Option 1 assumes that the work has been done to make attributes
+versionable and consistent with
+[#4](https://github.com/TheFoundryVisionmongers/OpenAssetIO/issues/4),
+and that the relevant API methods have been updated to work with trait
+sets.
+
+### Preamble
+
+The examples assume the following traits and specifications are
+available - allowing a file-based image to be described:
+
+```cpp
+trait BlobTrait {
+    name: "blob",
+    props: { url: str, mimeType: str }
+}
+
+trait ImageTrait {
+    name: "image",
+    props: { colorSpace: str }
+}
+
+spec ImageSpec {
+    traits: ["blob", "image"]
+}
+```
+
+### Publishing an image
+
+This example illustrates a host wanting to publish an image with its
+associated color space, to ensure it is accurately tracked.
+
+```python
+img = ImageSpec()
```

Comment:
Do you mind leaving this as-is? I quite liked the fact that it tied in more concretely with the declaration of the spec in the preamble, reenforcing the "robustness", even if it is at the expense of accuracy to the case we performance tested. (It would be better right, as the trait IDs would be in a static vector rather than on the stack?)